### ISSUE TITLE
Blockchain Engineering Class of 2020: Delft-DAO - Freedom of Computing

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#5146]
### Delft-DAO
What is a DAO? Within the coming decade the future of jobs, employment and the nature of the firm will change profoundly. Automation, AI, and robots will replace many of today's jobs. A new type of company is a company without any employees, without any machines or physical infrastructure. A Decentralized Autonomous Organizations, DAO, only exists in software. It goes beyond smart contracts, it is a complete company inside software. DAO development is still in the experimental stage.

Within this course you can create your very own autonomous organisation, the Delft-DAO. Learn to engineer a decentralised autonomous organisation, use the existing tools, and understand the security risks. The aim is to alter the nature of the firm in the Internet age, [see the Nobel prize winning theory](https://en.wikipedia.org/wiki/The_Nature_of_the_Firm). Production cost become essentially cost-free. An organisation which exists purely in cyberspace. The Delft-DAO is designed to be the first sustainable DAO. How can we empower leaderless organizations? How can it earn money from manipulating bits? 

Several teams from the class will create running code (requirements for a passing grade..) for one of the following functionalities. Four teams together create the building blocks of the complete Delft-DAO (time consuming integration step is outside our scope).

**Freedom of Computing**
Freely executing code forms the core of the DAO. Improvement of smart contract execution model. This team will build the enabling infrastructure of the DAO. You will create Python-based code execution using the "Dapp" approach, decentralized applications. All executed code on the Delft-DAO is compressed down to a single hash. This code can be upgraded and [can never revoked in a "rage-quit" incident](https://arstechnica.com/information-technology/2016/03/rage-quit-coder-unpublished-17-lines-of-javascript-and-broke-the-internet/); it probably is even "lawyer-proof". The [mandatory starting point for your code](https://github.com/mitchellolsthoorn/ipv8-module-loader) with documentation in [this master thesis](https://repository.tudelft.nl/islandora/object/uuid%3Ad68197ec-50b9-4452-b9bf-e34a743f165f).
  - create DAO computing layer
  - expand core with Python [sandboxing](https://wiki.python.org/moin/SandboxedPython)
  - live Delft-DAO source code patching
  - demonstrate ability do live code updates at all online nodes

Welcome to the course!
Sprint proposal divided on team of 4 students :
- DAO stuff: https://hackingdistributed.com/2016/06/18/analysis-of-the-dao-exploit/
- get familiar with [Trustchain concept read docs](https://github.com/Tribler/kotlin-ipv8/blob/master/doc/INDEX.md)
  - some code: https://spectrum.ieee.org/tech-talk/computing/networks/ethereums-150-million-dollar-dao-opens-for-business-just-as-researchers-call-for-a-moratorium
  - fantasy idea? https://medium.com/moatcoin/an-experiment-fund-managing-protocol-powered-by-ethereum-blockchain-with-dao-implemented-9aae19b243fd

- Coding sprint 1:
  - Study: ipv8-module-loader
  - Study context: https://github.com/Tribler/tribler/issues/2943
  - understand the signing of code
  - Create Hello World2 Dapp
  - Design&document on this issue for a DAO Dapp:
    - required functionality
    - core _must have_ feature suitable for next coding sprint

**Comment by synctext** [comment]
Welcome to the course!
Sprint proposal divided on team of 4 students :
- DAO stuff: https://hackingdistributed.com/2016/06/18/analysis-of-the-dao-exploit/
- get familiar with [Trustchain concept read docs](https://github.com/Tribler/kotlin-ipv8/blob/master/doc/INDEX.md)
  - some code: https://spectrum.ieee.org/tech-talk/computing/networks/ethereums-150-million-dollar-dao-opens-for-business-just-as-researchers-call-for-a-moratorium
  - fantasy idea? https://medium.com/moatcoin/an-experiment-fund-managing-protocol-powered-by-ethereum-blockchain-with-dao-implemented-9aae19b243fd

- Coding sprint 1:
  - Study: ipv8-module-loader
  - Study context: https://github.com/Tribler/tribler/issues/2943
  - understand the signing of code
  - Create Hello World2 Dapp
  - Create infrastructure to collectively control 1 Bitcoin
    - assume for now a single trustworthy person managing the private key
    - distribute the key (threshold?)
    - A majority is required to spend this coin

**Comment by synctext** [comment]
sprint advice:
- no running code yet
- discussed "FBase-mobile" native in Kotlin
- next sprint goal: _download source code from The Internet and run them on Android_
- Key accomplishment will be live code injection in the most simple "Hello World" way
  - Try Kotlin or Java.
  - you work against the Android security mechanism, be good Malware!
  - evaluate if this is solving everything: https://github.com/lukeFalsina/Grab-n-Run
  - Try to integrate offline compiler such as CXXDroid, CppDroid (ignore the dependency library import hell)
  - [offline Kotlin Programming Compiler](https://apkpure.com/kotlin-programming-compiler/com.krazeapps.kotlinprogrammingcompiler)
  - backup plan: use the Python interpreter or C service with native .
  - "Dynamic Delivery" is not usable because code is fixed at Google store upload time, not at runtime.

**Comment by synctext** [comment]
Addition: please focus this sprint primarily on **Grab 'N Run**. I think it offers exactly what we need. Separately, also test and evaluate [dynamic-load-apk](https://github.com/singwhatiwanna/dynamic-load-apk/blob/master/README-en.md). Use something as simple as possible.
From the [Grab 'N Run research paper](https://sites.cs.ucsb.edu/~vigna/publications/2015_ACSAC_GrabNRun.pdf)
```
This additional piece of code is loaded at runtime (thus the
term “dynamic”), and it might not necessarily be present
in the application package at installation time: In fact, the
additional code could be retrieved, once again at runtime,
from a remote network endpoint.
```
Try to get this to work or create an issue on their repo what is broken.
Future sprint options:
- use Libtorrent to download executable code
- Realise the source code distribution model of FBase. Create an offline Kotlin compiler in a stand-alone .APK that can create installable apps on the device. See prior fancy _self-compile_ [running code from TUDelft](https://github.com/Tribler/self-compile-Android).

**Comment by Thystler** [comment]
@rootmonkey

**Comment by rootmonkey** [comment]
Sprint update:

- After some research, we found that Grab N Run internally uses DexClassLoader to securely and dynamically load classes, thus, instead of using their implementation we just tried and got DexClassLoader to load a class from a jar file. Currently our code just runs on an Android 10.0 (API 29) emulator as we tried to run the code on a Samsung device and we failed. So our guess is maybe we need a device with stock android on it. 

- Below we have a simple depiction of our running code: 
Before:
<img width="380" alt="Screenshot 2020-03-11 at 23 25 50" src="https://user-images.githubusercontent.com/48287237/76470051-f3fd0380-63ef-11ea-974c-55082e0791ac.png">

After button press:
<img width="380" alt="Screenshot 2020-03-11 at 23 26 17" src="https://user-images.githubusercontent.com/48287237/76470070-ff502f00-63ef-11ea-976b-5a721aedb26b.png">

Where the modified string is dynamically loaded from the method get_str of this class at runtime:
<img width="379" alt="Screenshot 2020-03-11 at 23 26 34" src="https://user-images.githubusercontent.com/48287237/76470213-155def80-63f0-11ea-8508-5d94239bf94b.png">

- While researching stuff about DexClassLoader we found this paper: https://sites.cs.ucsb.edu/~chris/research/doc/ndss14_execute.pdf which describes various ways dynamically load code into a running application. One of them was using java Runtime class which has a function Runtime.exec() which works similarly to C function system() for running shell commands. By using Runtime.exec() we can also execute another file but for that we need root privileges.

**Comment by tpapak01** [comment]
Update: We managed to get DexClassLoader to load a class from a jar file on an actual phone. The problem was that the application required runtime permissions through the actual device (at least, that's what solved a common problem we were experiencing). 
Since DexClassLoader eventually worked, and since Grab n' Run is based primarily on that class, we have high hopes that SecDexClassLoader will work as well. After all, the difficulty was the whole permission granting process on the actual device, and we achieved that without resorting to older versions and tools.
![Inject](https://user-images.githubusercontent.com/15090265/76475943-c2d8ff00-6400-11ea-8e03-eab762d705f0.PNG)
![Code](https://user-images.githubusercontent.com/15090265/76475946-c66c8600-6400-11ea-8e65-abb9139a8dae.PNG)
![Result](https://user-images.githubusercontent.com/15090265/76475960-cbc9d080-6400-11ea-8f40-f8181a71308a.PNG)

**Comment by synctext** [comment]
Sprint feedback:
- it works :smile:  (Android versions supported?)
- 40% of the course is now integration with the other teams, [see here](https://github.com/Tribler/tribler/issues/4863#issuecomment-596526327)
Next sprint focus:
- Clean import without any extra Android storage access permission using magnet links
- integrate Libtorrent and re-base on [Superapp](https://github.com/Tribler/trustchain-superapp/)
- strip the Grab 'N run or try to do this better (e.g. Delft way:-)
- Key problem is the GUI of executable code: re-use the sophisticated Webview solution of FBase

Wrap-up above sprint content, then demo to other groups at end of _week 6_ of this course?

**Comment by tpapak01** [comment]
We managed to get libtorrent working on android, whether a magnet link is used, or a .torrent file. Although the state of each download holds between runs (which means a download at 15% will continue from 15% the next time), we still haven't managed to find the actual file on the phone once the download is complete. Next steps:
- Locate downloaded file
- Create our own torrent referring to a dummy file and download that dummy file
- Create our own torrent referring to a .jar executable code file, so that we can take it "in-code" after the download and execute it using Dexclassloader
The following zip is a video of "working code".

[Libtorrent using both torrent and magnet link.zip](https://github.com/Tribler/tribler/files/4345939/Libtorrent.using.both.torrent.and.magnet.link.zip)

**Comment by synctext** [comment]
sprint progress:
- stateless issue and make code imports permanent
- no communication yet with other teams :cry: 
- unit testing :sob: 
- Libtorrent is downloading :+1: 
- superapp integration planned

**Comment by tpapak01** [comment]
Integration with superapp (as a separate module) is complete. GUI screenshots will follow soon.

![Screenshot_20200325-012236](https://user-images.githubusercontent.com/15090265/77490844-67fdc980-6e3c-11ea-8ece-839a2a4f0f4c.png)

I guess one main question about the integration with the other teams is: "what are we supposed to transfer from peer to peer in the form of torrents? Is it votes? Is it executable code, and if so, for what purpose? The other teams made shared wallets, methods of money transferring and a system to vote/agree on money transferring. Where does our contribution fit in this?"

**Comment by synctext** [comment]
Sprint feedback:
- [ ] smart contract demonstrator, based on FBase principles
- [ ] what is the Delft-DAO smart contract _demo_

From prior list:
- [ ] stateless issue and make code imports persistent (try to do it with no extra Android permissions)
- [ ] sophisticated Webview solution of FBase still on ToDo list

**Comment by synctext** [comment]
Sprint progress:
- [ ] Focus on basic integration and "Hello World" demo for other teams (first priority)
- [ ] module inside superapp (```Create a pull request of your branch into the dao branch of Tribler/trustchain-superapp repository```)
- [ ] define the structure of the APK: ipv8 community loading, put state in local sql, trustchain DB, or another design approach
- [ ] State preservation is not yet finished (2nd priority)
- [ ] integrated Webview in the Superapp not yet finished (2nd priority)

**Comment by tpapak01** [comment]
- Torrents can now be created, seeded and downloaded from peer to peer within the app 
- Made good progress on the state-persistence matter

**Comment by synctext** [comment]
Sprint progress:
- Freedom of computing "Core" progress
  - DexClassLoader only loads a specific class or method
  - DexClassLoader does **not load** GUI, resources, .XML files, layout specification, and various assets in general
  - Load .APK the China-way: https://github.com/didi/VirtualAPK
    - investigate their "trick"
    - document in a few lines and replicate?
  - Grab N Run internally uses DexClassLoader

**Comment by Thystler** [comment]
magnet:?xt=urn:btih:6d651f45a01afc35572c734c73cbf8a8f82c396f&dn=20+Awesome+Nature+Ultra+HD+4K+Wallpapers+3840x2160+Px+%5BSet+64%5D&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969&tr=udp%3A%2F%2Ftracker.openbittorrent.com%3A80&tr=udp%3A%2F%2Fopen.demonii.com%3A1337&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969&tr=udp%3A%2F%2Fexodus.desync.com%3A6969

**Comment by synctext** [comment]
Sprint progress:
- We need a GUI which can be loaded through a ```DEXclassLoader()```
- Otherwise you need to load other assets, XMLs, and isolated files
- Sected solution: [Tutorial: dynamic user interfaces in Android with fragments](https://www.vogella.com/tutorials/AndroidFragments/article.html) and the official [docs](https://developer.android.com/training/basics/fragments/fragment-ui)

**Comment by rootmonkey** [comment]
This was one of the basic test where the button is made in a different apk and loaded dynamically into Test4

<img width="377" alt="Screenshot 2020-04-15 at 15 49 04" src="https://user-images.githubusercontent.com/48287237/79344892-b03a7400-7f30-11ea-930c-e6b58ffa79f6.png">

**Comment by synctext** [comment]
@InvictusRMC from the MultiSIG wallet team has volunteered to help integration your efforts. Thus real Bitcoin bounty payout for a magnet link with useful code.
- developers seeds code in Libtorrent, announces private wallet, and sets bounty
- DAO members hold vote for approval
- when the quorum is reached, payout is triggered
- everybody happy :partying_face:

**Comment by tpapak01** [comment]
Our app/module is complete. Below you can find a guide describing the main use case, from start to finish, and each step required along the process:
https://github.com/tpapak01/trustchain-superapp/wiki/FOC---app
Here is the apk:
[app-debug.zip](https://github.com/Tribler/tribler/files/4519110/app-debug.zip)
We can test the functionality tomorrow, if you want to.
A new pull request was made for the superapp.
Finally, we spoke with the Democracy group, and agreed on the integration. In short, if a user wants to distribute code among the network (through our Upload button) he must first propose that code/apk and have it approved/voted for, otherwise the button blocks him from uploading his apk.

**Comment by synctext** [comment]
Final sprint feedback:
- only minor polish needed, good stuff!
- only 1 freeze and 1 crash :clap: It actually worked.
- GUI is a bit primitive

**Comment by synctext** [comment]
Testing latest release...
![Screenshot_20200430_115548_nl tudelft trustchain](https://user-images.githubusercontent.com/325224/80698047-367eba80-8ada-11ea-80d2-f54d27424b24.jpg)
